src/ucs/datastruct/mpool_set.c

```diff
@@ -0,0 +1,153 @@
+        }
+
+        mp_set->bitmap |= sizes[i];
+    }
+
+    /* max_mp_entry_size is allowed to be non pow of 2. If this is the case, set
+     * the last bit in the map indicating that max_mp_entry_size is not equal to
+     * any element in sizes array.
+     */
+    if (!ucs_is_pow2(max_mp_entry_size) ||
+        !(max_mp_entry_size & mp_set->bitmap)) {
+        mp_set->bitmap |= UCS_BIT(max_idx);
+    }
+
+    mpools_num     = ucs_popcount(mp_set->bitmap);
+    mp_set->mpools = ucs_malloc(mpools_num * sizeof(*mp_set->mpools) +
+                                    priv_size,
+                                "mpools_set");
+    if (mp_set->mpools == NULL) {
+        ucs_error("failed to allocate mpools set");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    mps_idx  = 0;
+    prev_idx = max_idx;
+    ucs_for_each_bit(size_log2, mp_set->bitmap) {
+        map_idx = max_idx - size_log2;
+        size    = (map_idx == 0) ? max_mp_entry_size : UCS_BIT(size_log2);
+        status  = ucs_mpool_init(&mp_set->mpools[mps_idx], priv_size,
+                                 size + priv_elem_size, align_offset, alignment,
+                                 elems_per_chunk, max_elems, ops, name);
+        if (status != UCS_OK) {
+            goto err;
+        }
+
+        /* mp_set->map is an array of pointers to memory pools. Array index is
+         * log2 of the corresponding memory pool element size. Indexation is
+         * done in the reverse order to speedup array index calculation in
+         * ucs_mpool_set_get(). Since there is no separate memory pool for
+         * every pow of 2 value, initialize all array elements in the range of
+         * [current_index, last_initalized_index).
+         * So, eventually every mp_set->map array element will point to a
+         * certain mpool with minimal element size capable of storing elements
+         * up to 2^(max_idx - index).
+         */
+        for (i = prev_idx; i >= map_idx; --i) {
+            mp_set->map[i] = &mp_set->mpools[mps_idx];
+        }
+        prev_idx = map_idx - 1;
+        ++mps_idx;
+    }
+
+    ucs_debug("mpool_set:%s, sizes map 0x%x, largest size %zu, mpools num %d",
+              name, mp_set->bitmap, max_mp_entry_size, mpools_num);
+
+    return UCS_OK;
+
+err:
+    for (i = 0; i < mps_idx; ++i) {
+        ucs_mpool_cleanup(&mp_set->mpools[i], 0);
+    }
+    ucs_free(mp_set->mpools);
+
+    return status;
+}
+
+void ucs_mpool_set_cleanup(ucs_mpool_set_t *mp_set, int leak_check)
+{
+    int i;
+
+    for (i = 0; i < ucs_popcount(mp_set->bitmap); ++i) {
+        ucs_mpool_cleanup(&mp_set->mpools[i], leak_check);
+    }
+
+    ucs_free(mp_set->mpools);
+}
+
+void *ucs_mpool_set_get(ucs_mpool_set_t *mp_set, size_t size)
+{
+    return ucs_mpool_set_get_inline(mp_set, size);
+}
+
+void ucs_mpool_set_put(void *obj)
+{
+    ucs_mpool_set_put_inline(obj);
+}
+
+void *ucs_mpool_set_priv(ucs_mpool_set_t *mp_set)
+{
+    return mp_set->mpools + ucs_popcount(mp_set->bitmap);
```

Comment:
how is "priv" used by UCP?

Reply:
it is not used by UCP, but it is used by UCT TX mpool in `uct_iface_mpool_init`

Reply:
ok, ic